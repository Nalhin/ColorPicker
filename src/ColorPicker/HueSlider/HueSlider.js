import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { SELECTOR_HEIGHT, SELECTED_SLIDER } from '../constants';
import ColorSelector from '../SliderHandle';

const Background = styled.div`
  display: inline-block;
  height: 10px;
  border-radius: 4px;
  width: 100%;
  background: linear-gradient(
    to right,
    hsl(0, 100%, 50%),
    hsl(60, 100%, 50%),
    hsl(120, 100%, 50%),
    hsl(180, 100%, 50%),
    hsl(240, 100%, 50%),
    hsl(300, 100%, 50%),
    hsl(360, 100%, 50%)
  );
`;

const HueSlider = ({
  selected,
  changeSelected,
  hueRef,
  hueColor,
  huePosition,
  changeHuePosition,
}) => {
  const getPosition = React.useCallback(
    event => {
      const slider = hueRef.current;
      const rect = slider.getBoundingClientRect();
      const x = Math.floor(
        (event.clientX !== undefined
          ? event.clientX
          : event.touches[0].clientX) - rect.left,
      );
      if (x <= 0) changeHuePosition({ x: 0, y: SELECTOR_HEIGHT });
      else if (x >= slider.offsetWidth)
        changeHuePosition({ x: slider.offsetWidth, y: SELECTOR_HEIGHT });
      else changeHuePosition({ x, y: SELECTOR_HEIGHT });
    },
    [changeHuePosition, hueRef],
  );

  const getPositionMove = React.useCallback(
    event => {
      event.preventDefault();
      if (selected === SELECTED_SLIDER.HUE) {
        getPosition(event);
      }
    },
    [getPosition, selected],
  );

  const getPositionClick = event => {
    getPosition(event);
    changeSelected(SELECTED_SLIDER.HUE);
  };

  React.useEffect(() => {
    if (selected === SELECTED_SLIDER.HUE) {
      window.addEventListener('mousemove', getPositionMove);
      window.addEventListener('touchmove', getPositionMove);
    } else {
      window.removeEventListener('mousemove', getPositionMove);
      window.removeEventListener('touchmove', getPositionMove);
    }
    return () => {
      window.removeEventListener('mousemove', getPositionMove);
      window.removeEventListener('touchmove', getPositionMove);
    };
  }, [selected, getPositionMove]);

  return (
    <div>
      <ColorSelector
        color={hueColor}
        position={huePosition}
        click={getPositionClick}
      >
        <Background ref={hueRef} />
      </ColorSelector>
    </div>
  );
};

HueSlider.propTypes = {
  hueColor: PropTypes.arrayOf(PropTypes.number).isRequired,
  selected: PropTypes.number.isRequired,
  changeSelected: PropTypes.func.isRequired,
  hueRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
  ]).isRequired,
  huePosition: PropTypes.shape({ x: PropTypes.number, y: PropTypes.number })
    .isRequired,
  changeHuePosition: PropTypes.func.isRequired,
};

export default HueSlider;
