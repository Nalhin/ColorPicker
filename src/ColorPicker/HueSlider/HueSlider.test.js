import React from 'react';
import { render } from '@testing-library/react';

import HueSlider from './HueSlider';

describe('Hue Slider ', () => {
  const props = {
    selected: 1,
    changeSelected: () => {},
    hueRef: () => {},
    hueColor: [255, 240, 220, 1],
    huePosition: { x: 40, y: 50 },
    changeHuePosition: () => {},
  };
  it('Should render correctly', () => {
    const { container } = render(<HueSlider {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
