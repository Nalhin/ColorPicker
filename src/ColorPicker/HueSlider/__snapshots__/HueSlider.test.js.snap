// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hue Slider  Should render correctly 1`] = `
<div>
  <div
    class="css-cqd1um-SliderHandle__Wrapper e9y9zbe0"
  >
    <div
      class="css-gjeaqo-SliderHandle__Handle e9y9zbe1"
      style="top: 50px; left: 33px; background: rgb(255, 240, 220);"
    />
    <div
      class="css-1ok0unb-HueSlider__Background e11e8k5c0"
    />
  </div>
</div>
`;
